query_optimizer/rules/BottomUpRule.hpp

```diff
@@ -80,6 +81,14 @@ class BottomUpRule : public Rule<TreeType> {
    */
   virtual TreeNodePtr applyToNode(const TreeNodePtr &node) = 0;
 
+  /**
+   * @brief Initialization code to be used for each node.
```

Comment:
The initialization is only applied with the tree's root node but not "each node".
Other places look good. I'll do this minor revision and then merge.